rfcs/0000-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,75 @@
+
+## Procedure
+
+1. The maintainer will prepare a PR with their intended changes.
+2. The maintainer should test a sample of dependent derivations to ensure that their change is not unnecessarily or unintentionally breaking. (Example: Ensure that a critical output file was not forgotten) Note that sometimes it **is** necessary to break all dependent packages, and the maintainer is not required to avoid this.
+3. The maintainer will get the PR reviewed and approved.
+    1. It is **recommended but not required** to have some maintainers of dependent packages involved in the review to review if the breakage is justified.
+4. The maintainer will contact the maintainers of all dependent, broken packages (herein called sub-maintainers).
+    1. The sub-maintainers will be provided a link to the PR with the breaking changes as well as any context that will help them resolve the breakages.
+    2. The sub-maintainers are expected to update the derivations that they maintain within 7 days.
+        1. These changes should be merged to `master` following the regular nixpkgs PR process if they are backwards compatible. Otherwise they will be merged into the branch of the breaking PR.
+5. The maintainer must provide a grace period of at least 7 days from when the sub-maintainers were notified.
+6. The maintainer must merge `master` into their PR branch.
+7. The maintainer must mark all still-broken packages as [broken](https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes).
+8. The maintainer can now merge to `master`.
+
+This procedure should not result in a failing package build in `master` at any point.
+
+# Drawbacks
+
+This delays the merge of “core” derivations as the author needs to wait for sub-packages to be tested and possibly fixed up to the 7 day threshold.
+
+# Alternatives
+## Shorter grace period
+
+A shorter grace period allows maintainers to move more quickly but requires maintainers of defendant derivations to jump to action quickly which is not always possible for volunteer driven work. For example if a maintainer is on vacation it can’t be expected that they respond in a couple of days.
+
+## Longer grace period
+
+A longer grace period would give derivation maintainers more time to react to changes in dependencies and provide more time to search for replacement maintainers if the original maintainer has abandoned the derivation. The 7 day number was mostly arbitrary and can easily be changed in the future. A longer grace period would also result in more packages being fixed before the breaking change is shipped which would mean that end-users would experience less breakage on channels.
+
+# Unresolved questions
+## Critical Packages and Tests
+
+What if a breaking change breaks NixOS tests? There must be packages and tests so critical that we can not merge without them passing? In that case do we leave the PR open until fixed? or use the `staging` branch?
```

Comment:
What is the scope of the change we are talking about? Staging is not for testing or dumping changes in the hope they get fixed; one is expected to follow their change and fix regressions especially when it has reached staging-next.

There are important NixOS tests and tests that are pretty unimportant.

Reply:
What are your suggestions? Should we have a list of tests that can't be marked broken? In that case I think we would need to block the initial PR,